BR -> GCS report download error

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
<!-- You can run `tikv-server --version` -->
```
TiKV 
Release Version:   4.1.0-alpha
Git Commit Hash:   4b00298465da5162c7f62365939f918a63b3f900
Git Commit Branch: HEAD
UTC Build Time:    2020-04-20 04:01:52
Rust Version:      rustc 1.42.0-nightly (0de96d37f 2019-12-19)
Enable Features:   jemalloc portable sse protobuf-codec
Profile:           dist_release
```
### What operating system and CPU are you using?
<!-- If you're using Linux, you can run `cat /proc/cpuinfo` -->
run in docker container 
### Steps to reproduce
<!-- If possible, provide a recipe for reproducing the error. A complete runnable program is good. -->
use br to restore backup 
the br command 
```
br restore full --send-credentials-to-tikv=true --storage=gcs://test-gcp-jony/ --gcs.predefined-acl=private --pd=cluster-pd.cluster:2379
```
### What did you expect?
restore successful
### What did happened?
report error 
```
[2020/04/22 12:28:29.666 +00:00] [ERROR] [sst_importer.rs:121] ["download failed"] [err="Engine Engine(\"Corruption: Bad table magic number: expected 9863518390377041911, found 3474576814686811188 in /var/lib/tikv/import/.temp/cf4ef1b9-3fc5-4c1d-86bd-c3083c90bd2d_62_1_31.sst\")"] [meta="uuid: CF4EF1B93FC54C1D86BDC3083C90BD2D range { start: 7480000000000000FF3E5F720000000000FA end: 7480000000000000FF3E5F72FF } cf_name: \"write\" region_id: 62 region_epoch { conf_ver: 1 version: 31 }"]
```
the file in gcs is 22.74M but the download file is 473 bytes
```
-rw-r--r--    1 root     root           473 Apr 22 12:28 /var/lib/tikv/import/.temp/cf4ef1b9-3fc5-4c1d-86bd-c3083c90bd2d_62_1_31.sst
```

--- a/components/external_storage/src/gcs.rs
+++ b/components/external_storage/src/gcs.rs
@@ -14,10 +14,10 @@ use std::{
 
 use bytes::Bytes;
 use futures_util::{
+    future::{FutureExt, TryFutureExt},
     io::AsyncRead,
-    stream::{StreamExt, TryStreamExt},
+    stream::TryStreamExt,
 };
-use http::Method;
 use kvproto::backup::Gcs as Config;
 use reqwest::{Body, Client};
 use tame_gcs::{
@@ -107,25 +107,28 @@ impl GCSStorage {
             .or_io_error("failed to build request")
     }
 
-    fn convert_response(&self, res: reqwest::Response) -> Result<http::Response<Bytes>> {
+    async fn convert_response(
+        &self,
+        tag: &str,
+        res: reqwest::Response,
+    ) -> Result<http::Response<Bytes>> {
         let mut builder = http::Response::builder()
             .status(res.status())
             .version(res.version());
         for (key, value) in res.headers().iter() {
             builder = builder.header(key, value);
         }
-        // Use blocking IO, since conver_response is only used to read access token.
-        let content = block_on_external_io(res.bytes())
-            .map_err(|e| Error::new(ErrorKind::Other, format!("failed to read response: {}", e)))?;
-        builder.body(content).map_err(|e| {
-            Error::new(
-                ErrorKind::Other,
-                format!("failed to build response body: {}", e),
-            )
-        })
+        // convert_response is only used to read access token.
+        let content = res
+            .bytes()
+            .await
+            .or_io_error(format_args!("failed to read {} response", tag))?;
+        builder
+            .body(content)
+            .or_io_error(format_args!("failed to build {} response body", tag))
     }
 
-    fn set_auth(&self, req: &mut reqwest::Request, scope: tame_gcs::Scopes) -> Result<()> {
+    async fn set_auth(&self, req: &mut reqwest::Request, scope: tame_gcs::Scopes) -> Result<()> {
         let token_or_request = self
             .svc_access
             .get_token(&[scope])
@@ -137,12 +140,12 @@ impl GCSStorage {
                 scope_hash,
                 ..
             } => {
-                // Use blocking IO.
-                let res =
-                    block_on_external_io(self.client.execute(request.into())).map_err(|e| {
-                        Error::new(ErrorKind::Other, format!("request token failed: {}", e))
-                    })?;
-                let response = self.convert_response(res)?;
+                let res = self
+                    .client
+                    .execute(request.into())
+                    .await
+                    .or_io_error("request GCS access token failed")?;
+                let response = self.convert_response("GCS access token", res).await?;
 
                 self.svc_access
                     .parse_token_response(scope_hash, response)
@@ -159,7 +162,7 @@ impl GCSStorage {
         Ok(())
     }
 
-    fn make_request(
+    async fn make_request(
         &self,
         mut req: reqwest::Request,
         scope: tame_gcs::Scopes,
@@ -179,20 +182,18 @@ impl GCSStorage {
             }
         }
 
-        self.set_auth(&mut req, scope)?;
-        let response = block_on_external_io(self.client.execute(req))
+        self.set_auth(&mut req, scope).await?;
+        let response = self
+            .client
+            .execute(req)
+            .await
             .or_io_error("make GCS request failed")?;
         if !response.status().is_success() {
             let status = response.status();
-            let text = block_on_external_io(response.text()).map_err(|e| {
-                Error::new(
-                    ErrorKind::Other,
-                    format!(
-                        "request failed and failed to read error message, status: {}, error: {}",
-                        status, e
-                    ),
-                )
-            })?;
+            let text = response.text().await.or_io_error(format_args!(
+                "GCS request failed and failed to read error message, status: {}, error",
+                status
+            ))?;
             return Err(Error::new(
                 ErrorKind::Other,
                 format!("request failed. status: {}, text: {}", status, text),
@@ -263,18 +264,14 @@ impl ExternalStorage for GCSStorage {
             ..Default::default()
         });
         let req = Object::insert_multipart(&bucket, reader, content_length, &metadata, optional)
-            .map_err(|e| {
-                Error::new(
-                    ErrorKind::Other,
-                    format!("failed to create insert request: {}", e),
-                )
-            })?;
-        self.make_request(self.convert_request(req)?, tame_gcs::Scopes::ReadWrite)?;
-
+            .or_io_error("failed to create GCS insert request")?;
+        block_on_external_io(
+            self.make_request(self.convert_request(req)?, tame_gcs::Scopes::ReadWrite),
+        )?;
         Ok(())
     }
 
-    fn read(&self, name: &str) -> Box<dyn AsyncRead + Unpin> {
+    fn read(&self, name: &str) -> Box<dyn AsyncRead + Unpin + '_> {
         let bucket = self.config.bucket.clone();
         let name = self.maybe_prefix_key(name);
         debug!("read file from GCS storage"; "key" => %name);
@@ -293,22 +290,20 @@ impl ExternalStorage for GCSStorage {
         if let Err(e) = body.read_to_end(&mut body_content) {
             return GCSStorage::error_to_async_read(ErrorKind::Other, e);
         }
-        let response = match self.make_request(
-            http::Request::from_parts(parts, body_content).into(),
-            tame_gcs::Scopes::ReadOnly,
-        ) {
-            Ok(response) => response,
-            Err(e) => return GCSStorage::error_to_async_read(ErrorKind::Other, e),
-        };
+
         Box::new(
-            response
-                .bytes_stream()
-                .map(|result| {
-                    result.map_err(|e| {
-                        Error::new(ErrorKind::Other, format!("download from gcs error {}", e))
-                    })
+            self.make_request(
+                http::Request::from_parts(parts, body_content).into(),
+                tame_gcs::Scopes::ReadOnly,
+            )
+            .boxed() // this `.boxed()` pin the future.
+            .map_ok(|response| {
+                response.bytes_stream().map_err(|e| {
+                    Error::new(ErrorKind::Other, format!("download from gcs error {}", e))
                 })
-                .into_async_read(),
+            })
+            .try_flatten_stream()
+            .into_async_read(),
         )
     }
 }

--- a/components/external_storage/src/gcs.rs
+++ b/components/external_storage/src/gcs.rs
@@ -7,6 +7,7 @@ use super::{
 
 use std::{
     convert::TryInto,
+    fmt::Display,
     io::{Error, ErrorKind, Read, Result},
     sync::Arc,
 };
@@ -39,6 +40,28 @@ pub struct GCSStorage {
     client: Client,
 }
 
+trait ResultExt {
+    type Ok;
+
+    // Maps the error of this result as an `std::io::Error` with `Other` error
+    // kind.
+    fn or_io_error<D: Display>(self, msg: D) -> Result<Self::Ok>;
+
+    // Maps the error of this result as an `std::io::Error` with `InvalidInput`
+    // error kind.
+    fn or_invalid_input<D: Display>(self, msg: D) -> Result<Self::Ok>;
+}
+
+impl<T, E: Display> ResultExt for std::result::Result<T, E> {
+    type Ok = T;
+    fn or_io_error<D: Display>(self, msg: D) -> Result<T> {
+        self.map_err(|e| Error::new(ErrorKind::Other, format!("{}: {}", msg, e)))
+    }
+    fn or_invalid_input<D: Display>(self, msg: D) -> Result<T> {
+        self.map_err(|e| Error::new(ErrorKind::InvalidInput, format!("{}: {}", msg, e)))
+    }
+}
+
 impl GCSStorage {
     /// Create a new GCS storage for the given config.
     pub fn new(config: &Config) -> Result<GCSStorage> {
@@ -48,24 +71,13 @@ impl GCSStorage {
         if config.credentials_blob.is_empty() {
             return Err(Error::new(ErrorKind::InvalidInput, "missing credentials"));
         }
-        let svc_info = ServiceAccountInfo::deserialize(&config.credentials_blob).map_err(|e| {
-            Error::new(
-                ErrorKind::InvalidInput,
-                format!("invalid credentials_blob: {}", e),
-            )
-        })?;
-        let svc_access = ServiceAccountAccess::new(svc_info).map_err(|e| {
-            Error::new(
-                ErrorKind::InvalidInput,
-                format!("invalid credentials_blob {}", e),
-            )
-        })?;
-        let client = Client::builder().build().map_err(|e| {
-            Error::new(
-                ErrorKind::Other,
-                format!("unable to create reqwest client: {}", e),
-            )
-        })?;
+        let svc_info = ServiceAccountInfo::deserialize(&config.credentials_blob)
+            .or_invalid_input("invalid credentials_blob")?;
+        let svc_access =
+            ServiceAccountAccess::new(svc_info).or_invalid_input("invalid credentials_blob")?;
+        let client = Client::builder()
+            .build()
+            .or_io_error("unable to create reqwest client")?;
         Ok(GCSStorage {
             config: config.clone(),
             svc_access: Arc::new(svc_access),
@@ -85,23 +97,14 @@ impl GCSStorage {
         R: AsyncRead + Send + Unpin,
     {
         let uri = req.uri().to_string();
-        let builder = match req.method().clone() {
-            Method::GET => self.client.get(&uri),
-            Method::POST => self.client.post(&uri),
-            method => {
-                return Err(Error::new(
-                    ErrorKind::Other,
-                    format!("method unimplemented: {}", method.as_str()),
-                ))
-            }
-        };
-        Ok(builder
+        self.client
+            .request(req.method().clone(), &uri)
             .headers(req.headers().clone())
             .body(Body::wrap_stream(AsyncReadAsSyncStreamOfBytes::new(
                 req.into_body(),
             )))
             .build()
-            .map_err(|e| Error::new(ErrorKind::Other, format!("failed to build request: {}", e)))?)
+            .or_io_error("failed to build request")
     }
 
     fn convert_response(&self, res: reqwest::Response) -> Result<http::Response<Bytes>> {
@@ -126,7 +129,7 @@ impl GCSStorage {
         let token_or_request = self
             .svc_access
             .get_token(&[scope])
-            .map_err(|e| Error::new(ErrorKind::Other, format!("failed to get token: {}", e)))?;
+            .or_io_error("failed to get token")?;
         let token = match token_or_request {
             TokenOrRequest::Token(token) => token,
             TokenOrRequest::Request {
@@ -143,19 +146,14 @@ impl GCSStorage {
 
                 self.svc_access
                     .parse_token_response(scope_hash, response)
-                    .map_err(|e| {
-                        Error::new(
-                            ErrorKind::Other,
-                            format!("failed to parse token response: {}", e),
-                        )
-                    })?
+                    .or_io_error("failed to parse GCS token response")?
             }
         };
         req.headers_mut().insert(
             http::header::AUTHORIZATION,
-            token.try_into().map_err(|e| {
-                Error::new(ErrorKind::Other, format!("failed to set auth token: {}", e))
-            })?,
+            token
+                .try_into()
+                .or_io_error("failed to set GCS auth token")?,
         );
 
         Ok(())
@@ -175,12 +173,7 @@ impl GCSStorage {
                     *req.url_mut() = reqwest::Url::parse(
                         &[endpoint.trim_end_matches('/'), &url[hardcoded.len()..]].concat(),
                     )
-                    .map_err(|e| {
-                        Error::new(
-                            ErrorKind::InvalidInput,
-                            format!("invalid custom GCS endpoint: {}", e),
-                        )
-                    })?;
+                    .or_invalid_input("invalid custom GCS endpoint")?;
                     break;
                 }
             }
@@ -188,7 +181,7 @@ impl GCSStorage {
 
         self.set_auth(&mut req, scope)?;
         let response = block_on_external_io(self.client.execute(req))
-            .map_err(|e| Error::new(ErrorKind::Other, format!("make request fail: {}", e)))?;
+            .or_io_error("make GCS request failed")?;
         if !response.status().is_success() {
             let status = response.status();
             let text = block_on_external_io(response.text()).map_err(|e| {
@@ -238,12 +231,10 @@ impl ExternalStorage for GCSStorage {
             None
         } else {
             Some(
-                serde_json::from_str(&self.config.storage_class).map_err(|e| {
-                    Error::new(
-                        ErrorKind::InvalidInput,
-                        format!("invalid storage_class {}: {}", self.config.storage_class, e),
-                    )
-                })?,
+                serde_json::from_str(&self.config.storage_class).or_invalid_input(format_args!(
+                    "invalid storage_class {}",
+                    self.config.storage_class
+                ))?,
             )
         };
         // Convert manually since PredefinedAcl doesn't implement Deserialize.

--- a/components/external_storage/src/gcs.rs
+++ b/components/external_storage/src/gcs.rs
@@ -26,6 +26,11 @@ use tame_gcs::{
 };
 use tame_oauth::gcp::{ServiceAccountAccess, ServiceAccountInfo, TokenOrRequest};
 
+const HARDCODED_ENDPOINTS: &[&str] = &[
+    "https://www.googleapis.com/upload/storage/v1",
+    "https://www.googleapis.com/storage/v1",
+];
+
 // GCS compatible storage
 #[derive(Clone)]
 pub struct GCSStorage {
@@ -161,6 +166,26 @@ impl GCSStorage {
         mut req: reqwest::Request,
         scope: tame_gcs::Scopes,
     ) -> Result<reqwest::Response> {
+        // replace the hard-coded GCS endpoint by the custom one.
+        let endpoint = self.config.get_endpoint();
+        if !endpoint.is_empty() {
+            let url = req.url().as_str();
+            for hardcoded in HARDCODED_ENDPOINTS {
+                if url.starts_with(hardcoded) {
+                    *req.url_mut() = reqwest::Url::parse(
+                        &[endpoint.trim_end_matches('/'), &url[hardcoded.len()..]].concat(),
+                    )
+                    .map_err(|e| {
+                        Error::new(
+                            ErrorKind::InvalidInput,
+                            format!("invalid custom GCS endpoint: {}", e),
+                        )
+                    })?;
+                    break;
+                }
+            }
+        }
+
         self.set_auth(&mut req, scope)?;
         let response = block_on_external_io(self.client.execute(req))
             .map_err(|e| Error::new(ErrorKind::Other, format!("make request fail: {}", e)))?;

--- a/components/external_storage/src/util.rs
+++ b/components/external_storage/src/util.rs
@@ -57,6 +57,14 @@ pub fn error_stream(e: io::Error) -> impl Stream<Item = io::Result<Bytes>> + Unp
 }
 
 /// Runs a future on the current thread involving external storage.
+///
+/// # Caveat
+///
+/// This function must never be nested. The future invoked by
+/// `block_on_external_io` must never call `block_on_external_io` again itself,
+/// otherwise the executor's states may be disrupted.
+///
+/// This means the future must only use async functions.
 // FIXME: get rid of this function, so that futures_executor::block_on is sufficient.
 pub fn block_on_external_io<F: Future>(f: F) -> F::Output {
     // we need a Tokio runtime rather than futures_executor::block_on because

--- a/components/sst_importer/src/errors.rs
+++ b/components/sst_importer/src/errors.rs
@@ -87,7 +87,7 @@ quick_error! {
         }
         CannotReadExternalStorage(url: String, name: String, err: IoError) {
             cause(err)
-            display("Cannot read {}/{}", url, name)
+            display("Cannot read {}/{}: {}", url, name, err)
         }
         WrongKeyPrefix(what: &'static str, key: Vec<u8>, prefix: Vec<u8>) {
             display("\

--- a/components/sst_importer/src/sst_importer.rs
+++ b/components/sst_importer/src/sst_importer.rs
@@ -3,7 +3,7 @@
 use std::borrow::Cow;
 use std::fmt;
 use std::fs::{self, File, OpenOptions};
-use std::io::Write;
+use std::io::{self, Write};
 use std::ops::Bound;
 use std::path::{Path, PathBuf};
 use std::sync::Arc;
@@ -123,11 +123,11 @@ impl SSTImporter {
         );
         match self.do_download::<E>(meta, backend, name, rewrite_rule, speed_limiter, sst_writer) {
             Ok(r) => {
-                info!("download"; "meta" => ?meta, "url" => %url_of_backend(backend), "range" => ?r);
+                info!("download"; "meta" => ?meta, "name" => name, "range" => ?r);
                 Ok(r)
             }
             Err(e) => {
-                error!("download failed"; "meta" => ?meta, "url" => %url_of_backend(backend), "err" => %e);
+                error!("download failed"; "meta" => ?meta, "name" => name, "err" => %e);
                 Err(e)
             }
         }
@@ -163,8 +163,18 @@ impl SSTImporter {
                     Error::CannotReadExternalStorage(url.to_string(), name.to_owned(), e)
                 })?;
             if meta.length != 0 && meta.length != file_length {
-                let reason = format!("length {}, expect {}", file_length, meta.length);
-                return Err(Error::FileCorrupted(path.temp, reason));
+                let reason = format!(
+                    "downloaded size {}, expected {}, local path {}",
+                    file_length,
+                    meta.length,
+                    path.temp.display()
+                );
+                let reason = io::Error::new(io::ErrorKind::InvalidData, reason);
+                return Err(Error::CannotReadExternalStorage(
+                    url.to_string(),
+                    name.to_owned(),
+                    reason,
+                ));
             }
             IMPORTER_DOWNLOAD_BYTES.observe(file_length as _);
             OpenOptions::new()

--- a/components/sst_importer/src/sst_importer.rs
+++ b/components/sst_importer/src/sst_importer.rs
@@ -123,11 +123,11 @@ impl SSTImporter {
         );
         match self.do_download::<E>(meta, backend, name, rewrite_rule, speed_limiter, sst_writer) {
             Ok(r) => {
-                info!("download"; "meta" => ?meta, "range" => ?r);
+                info!("download"; "meta" => ?meta, "url" => %url_of_backend(backend), "range" => ?r);
                 Ok(r)
             }
             Err(e) => {
-                error!("download failed"; "meta" => ?meta, "err" => %e);
+                error!("download failed"; "meta" => ?meta, "url" => %url_of_backend(backend), "err" => %e);
                 Err(e)
             }
         }

